src/library/ebuild-backend.c

```diff
@@ -0,0 +1,335 @@
+static int ebuild_destroy_backend(void);
+
+backend ebuild_backend = {
+		kEbuildBackend,
+		ebuild_init_backend,
+		ebuild_load_list,
+		ebuild_destroy_backend,
+		/* list initialization */
+		{0, 0, NULL},
+};
+
+typedef struct contents {
+	char *md5;
+	char *path;
+} ebuildfiles;
+
+struct epkg {
+	char *cpv;
+	char *slot;
+	char *repo;
+	int files;
+	ebuildfiles *content;
+};
+
+
+char* remove_newline(char* string) {
+    int len = strlen(string);
+    if (len > 0 && string[len-1] == '\n') {
+        string[len-1] = '\0';
+    }
+    return string;
+}
+
+/*
+ * Portage stores data about installed packages in the VDB (/var/db/pkg/).
+ * We care about /var/db/pkg/category/package-version/CONTENTS
+ * which lists files and directories that are installed as part of a package 'merge'
+ * operation. All files are prefixed with 'obj' and are in the format:
+ * obj /path/to/file $(md5sum /path/to/file) $(date -r /path/to/file "+%s")
+ * e.g.
+ * obj /usr/bin/clamscan 3ade185bd024e29880e959e6ad187515 1693552964
+ */
+static int ebuild_load_list(const conf_t *conf) {
+	list_empty(&ebuild_backend.list);
+	struct _hash_record *hashtable = NULL;
+	struct _hash_record **hashtable_ptr = &hashtable;
+
+	DIR *vdbdir;
+	struct dirent *vdbdp;
+
+	if ((vdbdir = opendir("/var/db/pkg")) == NULL) {
+		msg(LOG_ERR, "Could not open /var/db/pkg");
+		return 1;
+	}
+
+	struct epkg *vdbpackages = NULL;
+	int i = 0;
+
+	/*
+	 * recurse through category/package-version/ dirs,
+	 * process CONTENTS (files, md5s), repository, SLOT,
+	 * store in epkg array
+	*/
+	while ((vdbdp = readdir(vdbdir)) != NULL) {
+
+		if (vdbdp->d_type == DT_DIR && strcmp(vdbdp->d_name, ".") != 0 &&
+				strcmp(vdbdp->d_name, "..") != 0) {
+
+			char *catdir;
+			if (asprintf(&catdir, "/var/db/pkg/%s", vdbdp->d_name) == -1) {
+				catdir = NULL;
+			}
+
+			msg(LOG_INFO, "Loading category %s", vdbdp->d_name);
+
+			if (catdir) {
+				DIR *cat;
+				struct dirent *catdp;
+				if ((cat = opendir(catdir)) == NULL) {
+					msg(LOG_ERR, "Could not open %s", catdir);
+					free(catdir);
+					continue;
+				}
+
+				while ((catdp = readdir(cat)) != NULL) {
+
+					if (catdp->d_type == DT_DIR && strcmp(catdp->d_name, ".") != 0 &&
+							strcmp(catdp->d_name, "..") != 0) {
+						char *pkgverdir;
+
+						if (asprintf(&pkgverdir, "%s/%s", catdir, catdp->d_name) == -1) {
+							pkgverdir = NULL;
+						}
+
+						msg(LOG_INFO, "Loading package %s/%s", vdbdp->d_name, catdp->d_name);
+						char *pkgrepo = NULL;
+						char *pkgslot = NULL;
+						int pkgfiles = 0;
+						ebuildfiles* pkgcontents = NULL;
+
+
+						if (pkgverdir) {
+							DIR *pkgver;
+							struct dirent *pkgverdp;
+
+							if ((pkgver = opendir(pkgverdir)) == NULL) {
+								msg(LOG_ERR, "Could not open %s", pkgverdir);
+								free(pkgverdir);
+								continue;
+							}
+
+							while ((pkgverdp = readdir(pkgver)) != NULL) {
+
+								// SLOT
+								if (pkgverdp->d_type == DT_REG &&
```

Comment:
Let's try hoist some of the really indented stuff into their own functions, it's quite hard to follow atm and I bet you can reuse a lot of it